torch/csrc/autograd/profiler.h

```diff
@@ -203,18 +203,63 @@ TORCH_API void mark(std::string name, bool include_cuda = true);
 TORCH_API void pushRange(std::string name);
 TORCH_API void popRange();
 
+struct TORCH_API FunctionCallContext {
+  explicit FunctionCallContext(Function* fn);
+  explicit FunctionCallContext(std::string name, int64_t sequence_nr = -1);
+  explicit FunctionCallContext(const char* name_ptr, int64_t sequence_nr = -1);
+
+  Function* func() const {
+    return fn_;
+  }
+
+  const char* name() const {
+    return name_ptr_;
+  }
+
+  int64_t seq_nr() const {
+    return sequence_nr_;
+  }
+
+ private:
+  Function* fn_ = nullptr;
+  std::unique_ptr<std::string> owned_name_;
+  const char* name_ptr_ = nullptr;
+  int64_t sequence_nr_ = -1;
+  //TODO: inputs
+};
+
+using UpdateCallContextCallback = std::function<void(FunctionCallContext&)>;
+using RecordFunctionCallback = std::function<void(const FunctionCallContext&)>;
+struct RecordCallbacks {
+  std::vector<RecordFunctionCallback> start_callbacks;
+  std::vector<RecordFunctionCallback> end_callbacks;
+};
+
+// all calls to registerCallbacks and clearCallbacks
+// should be finished before using profiler
+TORCH_API void registerCallbacks(
+    const RecordFunctionCallback& start, const RecordFunctionCallback& end);
+TORCH_API void clearCallbacks();
+
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn);
+  explicit RecordFunction(Function* fn, UpdateCallContextCallback* ucb = nullptr);
```

Comment:
What are `UpdateCallContextFallback`s for? Why can't they be just implemented as `RecordFunctionCallback`s? This seems really complicated.

Reply:
UpdateCallContextCallback is used to pass extra data from the callsite and initialize additional fields of FunctionCallContext that are very expensive to do in general case

Reply:
Even with this comment, I do not understand how UpdateCallContextCallback will be used. There doesn't seem to be any example uses in the tests either.

Reply:
added an example (invokeTestRecordFunction) into test